src/main/java/htsjdk/samtools/util/AbstractRecordAndOffset.java

```diff
@@ -0,0 +1,164 @@
+ * one base for <code>SamLocusIterator.RecordAndOffset</code> implementation or one alignment block of
+ * <code>SAMRecord</code> for <code>TypedRecordAndOffset</code> implementation.
+ */
+public class AbstractRecordAndOffset {
+
+    /**
+     * A SAMRecord aligned to reference position
+     */
+    private final SAMRecord record;
+    /**
+     * Zero-based offset into the read corresponding to the current position in AbstractLocusInfo
+     */
+    private final int offset;
+    /**
+     * Length of alignment block of the read
+     */
+    private final int length;
+    /**
+     * A reference position to which read offset is aligned.
+     */
+    private final int refPos;
+    private int hash = 0;
+
+    /**
+     * @param record inner SAMRecord
+     * @param offset from the start of the read
+     * @param length of alignment block
+     * @param refPos corresponding to read offset reference position
+     */
+    public AbstractRecordAndOffset(final SAMRecord record, final int offset, int length, int refPos) {
+        validateIndex(offset, record.getBaseQualities());
+        if (length > record.getReadLength()) {
+            throw new IllegalArgumentException("Block length cannot be larger than whole read length");
+        }
+        this.offset = offset;
+        this.record = record;
+        this.length = length;
+        this.refPos = refPos;
+    }
+
+    /**
+     * @return offset of aligned read base from the start of the read.
+     */
+    public int getOffset() {
+        return offset;
+    }
+
+    /**
+     * @return inner <code>SAMRecord</code> object.
+     */
+    public SAMRecord getRecord() {
+        return record;
+    }
+
+    /**
+     * @return the read base according to <code>offset</code>.
+     */
+    public byte getReadBase() {
+        return record.getReadBases()[offset];
+    }
+
+    /**
+     * @return the base quality according to <code>offset</code>.
+     */
+    public byte getBaseQuality() {
+        return record.getBaseQualities()[offset];
+    }
+
+    /**
+     * @return the length of alignment block represented by the object.
+     */
+    public int getLength() {
+        return length;
+    }
+
+    /**
+     * @return the position in reference sequence, to which the start of alignment block is aligned.
+     */
+    public int getRefPos() {
+        return refPos;
+    }
+
+    /**
+     * @return read name of inner SAMRecord.
+     */
+    public String getReadName() {
+        return record.getReadName();
+    }
+
+    /**
+     * @return array of base qualities of inner SAMRecord.
+     */
+    public byte[] getBaseQualities() {
+        return record.getBaseQualities();
+    }
+
+
+    /**
+     * @param position in the reference
+     * @return base quality of a read base, corresponding to a given reference position
+     */
+    public byte getBaseQuality(int position) {
```

Comment:
This method is called by picard.analysis.FastWgsMetricsCollector.excludeByQuality() only (see changes in Picard)